### ISSUE TITLE
Verify generated Angular templates code follows the best practice

### ISSUE THREAD

**Issue by Tanya0609** [Microsoft/WebTemplateStudio#681]
Our team is trying to release Angular out of preview mode. We would like to have Angular experts and experienced users to provide some feedback on our generated Angular templates code to see if it follows the best practice. 

Some feedback has already been provided, if there are any more critical feedbacks or potential improvements, please feel free to add it in this issue.

**Comment by manekinekko** [comment]
Hi, here are few quick notes (from interal discussion with @Tanya0609 ):
- [x] the content of the src/constants.ts should be moves to the environment files under src/environments
- [x] the files under src/app/images should be moves to the assets folder src/assets and use the MatIconModule to register and use the SVG assets (see https://material.angular.io/components/icon/api)
- [x] you may wanna move the WarningComponent under src/app/shared/warning/ (to be created)
- [x] I'd rename src/app/components to src/app/app-shell
- [x] all folders and symbol names should be in lower kebab case, eg, Master_Details.module.ts --> master-details.moduke.ts
- [x] I'd probably move the interface definitions from the components to their related services, eg, IGridTextItem in src/app/components/Grid/GridBox/grid-box.component.ts should be moved to src/app/components/Grid/grid.service.ts
- [x] make sure you run the ng lint command to fix  all formatting issues (if any). You may wanna make also sure the app is building in prod mode.
- [x] We released Angular version 8 yesterday, you may wanna give it a shot 😉
- [x] I was also wondering if you are planning to add some tests (if that is relevant)?

Let me know if you need anything else.

**Comment by manekinekko** [comment]
Internal question from @Tanya0609 
> I have a question towards “- I'd rename src/app/components to src/app/app-shell”.  I used Angular CLI to generate an Angular app and use ng generate to generate more components. It seems each component is a folder under /app, rather than being group into another folder /app/app-shell or /app/components.

When you generate symbols using the CLI, by default everything will be generated at the root of `src/app` and the CLI will use the `app.module.ts` as the host module. We don't force any folder architecture in the CLI. 

The "app-shell" naming is just a suggestion I gave—sorry I should have made this explicit. I just noticed that all of your components and modules are all somehow related to the UI shell. "app-shell" is a common naming convention. But your are free to name it differently. 

If you want to specify a different folder, just use the desired folder name: `ng g c subfolder/component-name`. You can also provide other flags when generating symbols (see [angular.io/cli/generate](https://angular.io/cli/generate)).

Note: if the subfolder contains an NgModule, the CLI will add the newly created component to that module, otherwise it will lookup for the nearest parent NgModule. 

Lmk if that makes sense 😉

**Comment by Tanya0609** [comment]
More feedback from William @xWiiLLz under issue #668 #667 #666 #665

**Comment by jsondoo** [comment]
`ng test` also seems to be failing

**Comment by Tanya0609** [comment]
@manekinekko 
Regarding to "the content of the src/constants.ts should be moves to the environment files under src/environments", could you elaborate why it should be moved?

The content of the src/constants.ts are constants for Error-Message (related to HTTP request failure cases) and Endpoint. Those constants can be used by both dev environment and production environment (Correct me if I am wrong). Will it be good to leave them in one file? Or it will be more appropriate to move them to the environment files under src/environments?

**Comment by Tanya0609** [comment]
"Router modules should be separated from their modules, and the main routing-module should reflect all that's possible within the app. Currently, when looking up the AppModule's routing, it's impossible to know that there are other routes in the app. (Modules should also be lazy-loaded, since it isn't much work added and shows best practices in terms of bundling size)"

Had discussion with @xWiiLLz: Since our current generated code doesn't have a separate Module's routing file for each file and some of the components aren't properly generated (missing few files). Thus, we will address this suggestion along with file structure together.

- [x] Routing modules and file structure improvement

**Comment by Tanya0609** [comment]
@manekinekko 
We have improved based on the previous suggestions that we understood. This is our expected generate code for Angular. When you have time, we would like to get a second round feedback from you.

https://github.com/Tanya0609/angular/tree/master/test

**Comment by manekinekko** [comment]
# environment files

> Regarding to "the content of the src/constants.ts should be moves to the environment files under src/environments", could you elaborate why it should be moved?

So `src/environments/environment.prod.ts` and `src/environments/environment.ts` are special files in angular. They are used by the Angular CLI when serving or building your app. They are supposed to host any environment-related configuration, like endpoints or api keys...etc This is because usually, we may have different endpoints (or other configs) depending on whether you are in a dev env or prod env (or any envs in between, ie, stagging, pre-prod...)

So according to this, I think that this piece of configuration under `src/constants.ts` should live under `src/environments/environment.prod.ts` and `src/environments/environment.ts`:
```
ENDPOINT: {
    LIST : '/api/list',
    GRID : '/api/grid',
    MASTERDETAIL : '/api/masterdetail',
}
```
It would look like this:

```
// src/environments/environment.ts
export const environment = {
  production: false,
  endpoint: {
    list : '/api/list',
    grid : '/api/grid',
    masterdetail : '/api/masterdetail',
  }
};
```

And (they are the same):

```
// src/environments/environment.prod.ts
export const environment = {
  production: true,
  endpoint: {
    list : '/api/list',
    grid : '/api/grid',
    masterdetail : '/api/masterdetail',
  }
};
```

However, you are also using the `src/constants.ts` to host error messages. So this is usually done by implementing the i18n feature of Angular (https://angular.io/guide/i18n). But I think this might be too overkill for this project unless you are planning to support i18n in WTS? 

So here is my recommendation for this:
1. Move these constants to src/environments/environment.*.ts` (you will have to duplicate them)
2. Don't use constants for error messages and just hard code the messages in their components (giving the fact that the rest of your text and description are in plain English).

# symbol imports

I've noticed that symbol imports aren't consistent, some are relative and others are going through TS aliases, ie.:

```
// this is using TS aliases
import { WarningMessageModule } from 'src/app/shared/warning-message/warning-message.module';

// this is a relative import
import { MasterDetailPageComponent } from './master-detail-page/master-detail-page.component';
```

Both are valid, just choose one ^^

# app-shell

## app-shell.module.ts

I think that all modules under `src/app/app-shell` could be grouped in a newly created `src/app/app-shell/app-shell/app-shell.module.ts` in order to abstract away the whole UI app shell components, ie. : 

```
import { NgModule } from '@angular/core';
import { CommonModule } from '@angular/common';
import { BlankModule } from './blank/blank.module';
import { GridModule } from './grid/grid.module';
import { ListModule } from './list/list.module';

@NgModule({
  declarations: [],
  imports: [
    BlankModule,
    GridModule,
    ListModule,
    ...
    CommonModule
  ]
})
export class AppShellModule { }

```

## HeaderComponent

The class name `HeaderComponent` and component selector `app-header` don't match:

```
// src/app/app-shell/nav-bar/nav-bar.component.ts
@Component({
  selector: 'app-header',
  templateUrl: './nav-bar.component.html',
  styleUrls: ['./nav-bar.component.css']
})
export class HeaderComponent implements OnInit { }
```

It should be: `NavBarComponent` for the class name and `app-nav-bar` for the  component selector (or any other name, but they have to be consistent):

```
// src/app/app-shell/nav-bar/nav-bar.component.ts
@Component({
  selector: 'app-nav-bar',
  templateUrl: './nav-bar.component.html',
  styleUrls: ['./nav-bar.component.css']
})
export class NavBarComponent implements OnInit { }
```

### Bootstrap vs Material

I see you are using both the Bootstrap and Material UI frameworks:

```
@import '~bootstrap/dist/css/bootstrap.min.css';
@import '~@angular/material/prebuilt-themes/indigo-pink.css';
```

I guess you are using Bootstrap only for the layout, right? 

I'd recommend you stick with one or the other, they are basically the same. However, if you decide to choose Bootstrap, and because Bootstrap is not component-based, you'd probably wanna switch to ng-bootstrap, which is just Bootstrap but for Angular, or React-Bootstrap for React. This also would probably help to avoid CSS overrides:

```
  button:focus,
  button:active {
    outline: none !important;
    box-shadow: none !important;
  }
```

FYI, Material has also support for layout: https://material.angular.io/cdk/layout/overview

# build prod

The app isn't building in prod mode (`ng build --prod`) and throws this error (mostly typing issues):

```
ERROR in src/app/app-shell/master-detail/master-detail-page/master-detail-page.component.html(23,8): : Property 'status' does not exist on type 'IMasterDetailText'.
src/app/app-shell/master-detail/master-detail-page/master-detail-page.component.html(25,8): : Property 'orderDate' does not exist on type 'IMasterDetailText'.
src/app/app-shell/master-detail/master-detail-page/master-detail-page.component.html(27,8): : Property 'shipTo' does not exist on type 'IMasterDetailText'.
src/app/app-shell/master-detail/master-detail-page/master-detail-page.component.html(29,8): : Property 'orderTotal' does not exist on type 'IMasterDetailText'.
src/app/app-shell/master-detail/master-detail-page/master-detail-page.component.html(31,8): : Property 'longDescription' does not exist on type 'IMasterDetailText'.
```

# tests

`ng test` is throwing this error:

```
ERROR in src/app/app-shell/list/list-item/list-item.component.spec.ts(3,35): error TS2307: Cannot find module './list-item.component'.
```

# other

I have noticed that some files contain the Google license, is that intentional?

```
/*
Copyright Google LLC. All Rights Reserved.
Use of this source code is governed by an MIT-style license that
can be found in the LICENSE file at http://angular.io/license
*/
```


That's all for now :)
/cc @Tanya0609

**Comment by SahilTara** [comment]
>@import 'bootstrap/dist/css/bootstrap.min.css';
@import '@angular/material/prebuilt-themes/indigo-pink.css';

Hmm material shouldn't be there anymore to be honest. @Tanya0609 
Feel like this is a remnant of the days we had material instead of bootstrap.

> However, if you decide to choose Bootstrap, and because Bootstrap is not component-based, you'd probably wanna switch to ng-bootstrap, which is just Bootstrap but for Angular, or React-Bootstrap for React.

The reason we didn't use these component based libraries is because we wanted the user to not have to do too much to get rid of bootstrap and replace it with their own styling should they have to. If we used the bootstrap components I feel that the user would have more trouble changing the styling. 

Is there a way to use bootstrap components while still making it easy for the user to swap to their own style? Or are we stuck with our current method?

**Comment by Tanya0609** [comment]
@manekinekko  For the app-shell module, do you think if we could leave it as a future enhancement? Since our target market is beginners and intermediate developers, adding another layer may add a little bit of complexcity.

**Comment by Tanya0609** [comment]
@SahilTara @manekinekko The reason I added the material css is because the MatIconModule I used for rendering svg images. When the module gets imported, it uses @angular/material package. Without using a material css, it gives a warning in the app. That was why I used the material css. I will retake a look at this one.

**Comment by manekinekko** [comment]
Hey @SahilTara 

> The reason we didn't use these component based libraries is because we wanted the user to not have to do too much to get rid of bootstrap and replace it with their own styling should they have to. If we used the bootstrap components I feel that the user would have more trouble changing the styling.

If you have a look at the HTML of some components, eg `test/src/app/app-shell/list/list.component.html`:
```
<main class="container" id="mainContent">
  <div class="row">
    <div class="col mt-5 p-0">
      <h3>Angular list Template</h3>
    </div>
    <div class="col-12 p-0">
```
Or in `test/src/app/app-shell/list/list-item/list-item.component.html` :
```
<div class="row">
  <div class="col-11">
    <p class="mt-3">{{text}}</p>
  </div>
  <div class="col-1">
```

Please keep in mind that all these layout and spacing CSS classes `row`, `col-11` ...etc are Bootstrap dependant. Even more problematic, you also have to have the right DOM hierarchy in order to have the correct result. So, if a user wants to get rid of bootstrap, this will probably be challenging for them since they will have to reimplement the whole layout in a different library.

> Is there a way to use bootstrap components while still making it easy for the user to swap to their own style?

This is usually not easy to do. Each library has it's own API and requirements. However... (read below)

> are we stuck with our current method?

### Styling vs Layout

#### Short term option (for the 1st release)

Not really if you separate the styling from the layout and positioning. What I mean is that you can use the Web platform CSS flex and Grid for the layout/position part (see tutorial [here](https://css-tricks.com/snippets/css/a-guide-to-flexbox/)). Then you can keep using bootstrap or MD for the general styling. But keep in mind that adding any additional library would increase the bundle size which is also something you may wanna consider checking.

#### Long-term

IDEA: You can probably provide a (fixed) set of themes based on popular CSS libraries (Bootstrap, MD, Foundation...) so the user can choose the type of library they are already using and we can generate the correct styling for them.

### SVG Icons

Hi @Tanya0609 

> do you think if we could leave it as a future enhancement? 

Of course, this is not mandatory for now. However, this is part of our official recommendations. See more details [https://angular.io/guide/feature-modules](https://angular.io/guide/feature-modules)

> adding another layer may add a little bit of complexcity.

I wouldn't worry about this too much. Angular developers are used to this kind of module encapsulations 👍

>  The reason I added the material css is because the MatIconModule I used for rendering svg images. 

Gotcha.

> When the module gets imported, it uses @angular/material package. Without using a material css, it gives a warning in the app. 

Do you remember what error was that?

### Bootsratp vs MD

Also, if you decide to choose Bootstrap over MD, I think you should completely drop the MD dependency and go full Bootstrap. Then use whatever Bootstrap recommendation to integrate SVG icons.

Please note that Bootstrap and MD provide virtually the same features, they are both CSS libraries. However, they implement two completely different design systems. So there is no need to use them both.

**Comment by Tanya0609** [comment]
@manekinekko  Thank you for the quick and very well explained response :)

I think we will stick with Boostrap framework and try to utilize ng-bootstrap library for the places it can be utilized for the 1st release. In the future enhancement, definitely it should be like what you pointed out in **long-term**.

The warning I got was:
![warning](https://user-images.githubusercontent.com/12874317/59644757-2a3c7980-9123-11e9-8651-bbb59bce145d.PNG)

**Comment by SahilTara** [comment]
@manekinekko Good catch on the layout stuff. Yeah I completely agree that we don't fully meet the goal of allowing it to be easy for the users. I think what you mentioned about using bootstrap for the styling and using flex box for the layout would be better for the user short term. 

Also having a selection type thing would be a nice edition, something like "Choose a style: material, bootstrap...". 

Hopefully this is implemented before the interns leave but might not be in scope to implement all of the styles. Maybe starting off with the ability to choose material or bootstrap would be something that could be implemented before they leave if they choose to implement it.

**Comment by xWiiLLz** [comment]
> ### Bootsratp vs MD
> Also, if you decide to choose Bootstrap over MD, I think you should completely drop the MD dependency and go full Bootstrap. Then use whatever Bootstrap recommendation to integrate SVG icons.
> 
> Please note that Bootstrap and MD provide virtually the same features, they are both CSS libraries. However, they implement two completely different design systems. So there is no need to use them both.

Great feedback. Definitely agree that you should stick to one styling library, and in the future let the user choose one from a list of options before generating the template.

If you're going to use bootstrap first, I'd really recommend using ng-bootstrap or ngx-bootstrap ([seems like ng-bootstrap is more used though](https://www.npmtrends.com/ngx-bootstrap-vs-@ng-bootstrap/ng-bootstrap)). If you ever want any of the dynamic components of bootstrap (collapse/dialogs/etc.), these packages are the only options to avoid importing jQuery, which is [_in my opinion_] the worst thing you could do to an Angular app.. 😅

In the future, I would definitely suggest making a version for material, since it is the recommended styling library to go with Angular ([and is also more used according to npm trends](https://www.npmtrends.com/@angular/material-vs-primeng-vs-ngx-bootstrap-vs-@ng-bootstrap/ng-bootstrap-vs-angular-foundation-vs-angular-foundation-6)).

**Comment by manekinekko** [comment]
@Tanya0609 I think you tagged the wrong person by mistake in https://github.com/microsoft/WebTemplateStudio/issues/681#issuecomment-502897522

Regarding your warning, you are right, it's because you included `MatIconModule` in the `grid.module.ts` module.

```
import { MatIconModule } from '@angular/material/icon';

@NgModule({
  imports: [
    MatIconModule,
  ]
})
export class GridModule { }
```

So the warning you are getting is a security check on our side. 

Now, since you are importing the material icon registry for **just** the SVG mapping, ie:

```
iconRegistry.addSvgIcon(
      'grey-box',
      sanitizer.bypassSecurityTrustResourceUrl('assets/GreyBox.svg'));
``` 

It looks to me that you are pulling the whole material design dependencies + the stylesheet just for the SVG mapping, right?

If that is the case, I'd recommend completely dropping MD for now and simply use your SVG icon with a plain old `img` HTML element, eg:

```
<img src="assets/GreyBox.svg" alt="Grey Box">
```

And then use Bootstrap to style it as a button.

Also, if you do this, remember to remove the `@angular/material` dep from your [package.json](https://github.com/Tanya0609/angular/blob/master/test/package.json#L12), otherwise  this will be included into your final bundle.

/cc @SahilTara

**Comment by Tanya0609** [comment]
@manekinekko sorry about I tagged a wrong person. Yes, I were using material design just for the SVG mapping, thus like you recommended, I dropped MD and removed the package from package.json in this newly PR #766. 

@xWiiLLz is another great intern from our program but working in a different project. He has been used Angular for few years. Thus, we got some help from him too. He also commented on the styling and suggested us to use ng-bootrap if we are going to use Boostrap only.

So now, the generated code only utilize Bootstrap and we utilized ng-bootstrap Alter component for styling the alter error message in our List page.

**Comment by Tanya0609** [comment]
Thank you all for the constructive feedbacks, detailed explanation and help! Our team really appriciate it! All the PRs to address all the suggestions should be go in by the end of today. I will post a new generated angular code again. If you have time feel free to check out. 

https://github.com/Tanya0609/angular/tree/master/test/src/app/app-shell

**Comment by manekinekko** [comment]
Hey @Tanya0609 I just came across the following file and noticed something about unit testing that can be improved.

Take this example for instance, from the `list.component.spec.ts`:

https://github.com/microsoft/WebTemplateStudio/blob/c269a3cb4e8e9b053e43301e8cb0162f0dd34579/templates/Web/Pages/Angular.List/src/app/app-shell/Param_SourceName_Kebab/list.component.spec.ts#L16-L30

In this test file, we are obviously testing the `ListComponent`. However, as you can see from the `declarations` and `imports` dependencies, we are loading way more dependencies that this test needs. Unit tests should import and load only the necessary dependencies for a particular component to execute. All children component (those called in the template) shouldn't be loaded and added to `TestBed`. Otherwise, we will end up trying to test these children components as well (that's not the purpose of `list.component.spec.ts`). These other children components will be tested in their respective spec files.

Of course, doing so will trigger a runtime error because the `ListComponent` can't be bootstrapped.   That's why Angular provides a mechanism that allows you to deal with these test scenarios:

```
TestBed.configureTestingModule({
      declarations: [
        ListComponent
      ],
      imports: [
        HttpClientTestingModule,
      ],
      schemas: [NO_ERRORS_SCHEMA]
    }).compileComponents();
``` 

You just need to add `schemas: [NO_ERRORS_SCHEMA]` to your TestBed module and remove all unnecessary dependencies. FYI, passing NO_ERRORS_SCHEMA tells the compiler not to error when encountering elements whose dependences (declarations) weren't provided/found.

I've also noticed this issue in other component tests as well :)

Cheers.

**Comment by Tanya0609** [comment]
Thank you for your great contributions! I will close this issue since all the feedback regarding to first release are addressed, and will log the long term suggestions into another issue.
